### ISSUE TITLE
Including generated Kubernetes definition in tooling layer causes very long processing time

### ISSUE THREAD

**Issue by cueckoo** [cue-lang/cue#391]
_Originally opened by @seh in https://github.com/cuelang/cue/issues/391_

### What version of CUE are you using (`cue version`)?

<pre>
cue version devel darwin/amd64
</pre>
Built via _go get_ on 16 May 2020 against commit 8fcefc84bf1d6868608beb7680a916f57181d9ed.

### Does this issue reproduce with the latest release?

Yes, insofar as this version is the latest, though not actually released.

### What did you do?

Run _cue get go k8s.io/api/storage/v1_ to generate CUE definitions for the Kubernetes "storage/v1" API group and version.

Write a few objects that embed the `v1.#StorageClass` definition, and serialize just one of them  as YAML using the `encoding/yaml` package in the tooling layer.

See the following _txtar_ file summarizing a small set of files that set up this arrangement:

<details>
<summary>CUE module using Kubernetes #StorageClass definition</summary>

**NOTE: CUE-generated Kubernetes definitions are omitted here for brevity.**
```txtar
-- cue.mod/module.cue --
module: "example.com"
-- kubernetes/common/storage.cue --
package common

import (
	"example.com/kubernetes"
)

#storage: {
	objects: [
		kubernetes.#StorageClass & {
			metadata: name: "sc1"
			provisioner: "something"
		},
		kubernetes.#StorageClass & {
			metadata: name: "sc2"
			provisioner: "something"
		},
		kubernetes.#StorageClass & {
			metadata: name: "sc3"
			provisioner: "something"
		},
		kubernetes.#StorageClass & {
			metadata: name: "sc4"
			provisioner: "something"
		},
		kubernetes.#StorageClass & {
			metadata: name: "sc5"
			provisioner: "something"
		},
	]
}
-- kubernetes/common/storage_tool.cue --
package common

import (
	"encoding/yaml"
	"tool/cli"
)

let sc = #storage.objects[0]

command: "test": {
	task: print: cli.Print & {
		text: yaml.Marshal(sc)
	}
}
-- kubernetes/storage.cue --
package kubernetes

import (
        storagev1 "k8s.io/api/storage/v1"
)

let baseObject = {
	apiVersion: "storage.k8s.io/v1"
}

#StorageClass: baseObject & {
	storagev1.#StorageClass

	kind: "StorageClass"
	volumeBindingMode: "WaitForFirstConsumer"
}
-- with-kubernetes-def.txtar --
```
</details>

### What did you expect to see?

Running _cue test 'example.com/kubernetes/common'_ would emit a single YAML document quickly, in a fraction of a second.

### What did you see instead?

Running _cue test 'example.com/kubernetes/common'_ takes approximately 70 seconds per CUE struct that embeds the `v1.#StorageClass` definition. I adjusted the entries in the `#storage.objects` list in file _kubernetes/common/storage.cue_ in each of these runs.

Object Count | Elapsed Wall Time (seconds)
--- | ---
1 |   67.15
2 | 140.20
3 | 212.71
4 | 277.97
5 | 346.80

If instead we omit the embedding of the `v1.#StorageClass` definition from file _kubernetes/storage.cue_, and define a few of the lost fields ourselves, we see the same output, but with dramatically lower processing time. Here is an amended _txtar_ file showing the files without that generated dependency:

<details>
<summary>CUE module omitting use of Kubernetes #StorageClass definition</summary>

```txtar
-- cue.mod/module.cue --
module: "example.com"
-- kubernetes/common/storage.cue --
package common

import (
	"example.com/kubernetes"
)

#storage: {
	objects: [
		kubernetes.#StorageClass & {
			metadata: name: "sc1"
			provisioner: "something"
		},
		kubernetes.#StorageClass & {
			metadata: name: "sc2"
			provisioner: "something"
		},
		kubernetes.#StorageClass & {
			metadata: name: "sc3"
			provisioner: "something"
		},
		kubernetes.#StorageClass & {
			metadata: name: "sc4"
			provisioner: "something"
		},
		kubernetes.#StorageClass & {
			metadata: name: "sc5"
			provisioner: "something"
		},
	]
}
-- kubernetes/common/storage_tool.cue --
package common

import (
	"encoding/yaml"
	"tool/cli"
)

let sc = #storage.objects[0]

command: "test": {
	task: print: cli.Print & {
		text: yaml.Marshal(sc)
	}
}
-- kubernetes/storage.cue --
package kubernetes

let baseObject = {
	apiVersion: "storage.k8s.io/v1"
}

#StorageClass: baseObject & {
	kind:              "StorageClass"
	volumeBindingMode: "WaitForFirstConsumer"

	metadata: name: !=""
	provisioner: !=""
}
-- sans-kubernetes-def.txtar --
```
</details>

Once again, if we now run _cue test 'example.com/kubernetes/common'_ with this modified file, it takes the same amount of time regardless of whether we emit one or five objects.

Object Count | Elapsed Wall Time (seconds)
--- | ---
1 |  0.04
2 | 0.04
3 | 0.04
4 | 0.04
5 | 0.04

This example is whittled down from a larger code base. Trying to emit just five objects there takes at least five minutes; I've rarely let it run long enough to see it complete. During that time, CUE burns around four CPUs and consumes approximately 9 GB of memory.

Note that I'm able to run commands like _cue eval_, _cue vet_, and _cue export_ against the non-tool files, and they complete almost immediately. Something about the split into the tooling layer for the YAML serialization triggers this behavior—even though in each of these tests we're only serializing **one** of the objects (per the "sc" let declaration in file _kubernetes/common/storage_tool.cue_), regardless of how many we define.

Why does my use of this generated definition perturb the processing time like this? Why does this processing time increase linearly with each object that embeds such a definition?

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/391#issuecomment-663978313_

The overlay layer is gone in v0.3.0-alpha1. I was not able to reproduce this with that version.

As a result there may be name clashes now, see release notes.

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/391#issuecomment-701487873_

This seems to have been fixed in v0.3.0:
 time cue test 'example.com/kubernetes/common' 
...
real    0m1.849s
user    0m3.928s
sys     0m0.118s

Up to v0.2.2: very long running times.

**Comment by cueckoo** [comment]
_Original reply by @seh in https://github.com/cuelang/cue/issues/391#issuecomment-701491812_

I confirmed similar timing with version _0.0.3-alpha3_ on macOS: `3.64s user 0.15s system 235%!c(MISSING)pu 1.615 total`.
Thank you!
